saq/queue/postgres.py

```diff
@@ -0,0 +1,608 @@
+        async with self.pool.connection() as conn, conn.cursor() as cursor:
+            if kind == "queued":
+                await cursor.execute(
+                    SQL(
+                        """
+                    SELECT count(*) FROM {jobs_table}
+                    WHERE status = 'queued'
+                      AND queue = %(queue)s
+                      AND %(now)s >= scheduled
+                    """
+                    ).format(jobs_table=self.jobs_table),
+                    {"queue": self.name, "now": math.ceil(seconds(now()))},
+                )
+            elif kind == "active":
+                await cursor.execute(
+                    SQL(
+                        """
+                    SELECT count(*) FROM {jobs_table}
+                    WHERE status = 'active'
+                      AND queue = %(queue)s
+                    """
+                    ).format(jobs_table=self.jobs_table),
+                    {"queue": self.name},
+                )
+            elif kind == "incomplete":
+                await cursor.execute(
+                    SQL(
+                        """
+                    SELECT count(*) FROM {jobs_table}
+                    WHERE status IN ('new', 'deferred', 'queued', 'active')
+                      AND queue = %(queue)s
+                    """
+                    ).format(jobs_table=self.jobs_table),
+                    {"queue": self.name},
+                )
+            else:
+                raise ValueError("Can't count unknown type {kind}")
+
+            result = await cursor.fetchone()
+            assert result
+            return result[0]
+
+    async def schedule(self, lock: int = 1) -> t.Any: ...
+
+    async def sweep(self, lock: int = 60, abort: float = 5.0) -> list[t.Any]:
+        """Delete jobs past their TTL from the jobs table"""
+        swept = []
+
+        async with self.pool.connection() as conn, conn.cursor() as cursor:
+            await cursor.execute(
+                """
+            SELECT pg_try_advisory_xact_lock(%(classid)s, %(objid)s)
+            """,
+                {"classid": SWEEP_LOCK_KEY[0], "objid": SWEEP_LOCK_KEY[1]},
+            )
+            result = await cursor.fetchone()
+            if result and not result[0]:
+                # Could not acquire the sweep lock so another worker must already be sweeping
+                return []
+            await cursor.execute(
+                SQL(
+                    """
+                DELETE FROM {jobs_table}
+                WHERE status IN ('aborted', 'complete', 'failed')
+                  AND %(now)s >= ttl
+                RETURNING key
+                """
+                ).format(jobs_table=self.jobs_table),
+                {"now": math.ceil(seconds(now()))},
+            )
+            await cursor.execute(
+                SQL(
+                    """
+                SELECT job
+                FROM {jobs_table} LEFT OUTER JOIN pg_locks ON lock_key = objid
```

Comment:
how does lock_key = objid if objid is null? null never == null

Reply:
Right, but there could be null objids because it's a left outer join if some jobs don't have locks.